src/NServiceBus.Core/Persistence/InMemory/SagaPersister/InMemorySagaPersister.cs

```diff
@@ -3,203 +3,218 @@ namespace NServiceBus
+
             return Task.FromResult(default(TSagaData));
         }
 
-        public Task<TSagaData> Get<TSagaData>(Guid sagaId, SynchronizedStorageSession session, ContextBag context) where TSagaData : IContainSagaData
+        public Task<TSagaData> Get<TSagaData>(string propertyName, object propertyValue, SynchronizedStorageSession session, ContextBag context) where TSagaData : IContainSagaData
         {
-            VersionedSagaEntity result;
-            if (data.TryGetValue(sagaId, out result) && result?.SagaData is TSagaData)
+            var key = GetKey(typeof(TSagaData), propertyName, propertyValue);
+            Guid id;
+
+            if (byCorrelationId.TryGetValue(key, out id))
             {
-                var sagaData = result.Read<TSagaData>();
-                return Task.FromResult(sagaData);
+                // this isn't updated atomically and may return null for an entry that has been indexed but not inserted yet
+                return Get<TSagaData>(id, session, context);
             }
+
             return Task.FromResult(default(TSagaData));
         }
 
         public Task Save(IContainSagaData sagaData, SagaCorrelationProperty correlationProperty, SynchronizedStorageSession session, ContextBag context)
         {
-            var inMemSession = (InMemorySynchronizedStorageSession) session;
-            inMemSession.Enlist(() =>
+            ((InMemorySynchronizedStorageSession)session).Enlist(() =>
             {
-                var lockenTokenKey = $"{sagaData.GetType().FullName}.{correlationProperty?.Name ?? "None"}.{correlationProperty?.Value ?? "None"}";
-                var lockToken = lockers.GetOrAdd(lockenTokenKey, key => new object());
-                lock (lockToken)
+                var correlationId = NoCorrelationId;
+                if (correlationProperty != SagaCorrelationProperty.None)
                 {
-                    if (correlationProperty != SagaCorrelationProperty.None)
+                    correlationId = GetKey(sagaData.GetType(), correlationProperty);
+                    if (byCorrelationId.TryAdd(correlationId, sagaData.Id) == false)
                     {
-                        ValidateUniqueProperties(correlationProperty, sagaData);
+                        throw new InvalidOperationException($"The saga with the correlation id 'Name: {correlationProperty.Name} Value: {correlationProperty.Value}' already exists");
                     }
+                }
 
-                    data.AddOrUpdate(sagaData.Id,
-                        id => new VersionedSagaEntity(sagaData, lockenTokenKey),
-                        (id, original) => new VersionedSagaEntity(sagaData, lockenTokenKey, original)); // we can never end up here.
+                var entry = new Entry(sagaData, correlationId);
+                if (sagas.TryAdd(sagaData.Id, entry) == false)
+                {
+                    throw new Exception("FATAL: this should never happened as saga id should be unique");
```

Comment:
What does `FATAL` even mean? Is there a better way to express the message in the exception why it should never happen?